### PR TITLE
Fix slice::ChunksMut aliasing

### LINKED ISSUE #94231: Stacked borrows fails on `{ChunksMut,ChunksExactMut}::__iterator_get_unchecked()`
### Description

```rust
fn main() {
    let mut arr1 = [0u8; 64];
    let arr2 = [0u8; 64];
    let mut iter = arr1.chunks_mut(8).zip(arr2.chunks(8));
    while let Some((chunk1, chunk2)) = iter.next() {
        dbg!(chunk2[0]);
        dbg!(chunk1[0]);
        iter.next();
        dbg!(chunk2[0]);
        dbg!(chunk1[0]);  // error here
    }
}
```

Running this code in Miri will produce the following output:

```
[src\main.rs:6] chunk2[0] = 0
[src\main.rs:7] chunk1[0] = 0
[src\main.rs:9] chunk2[0] = 0
error: Undefined Behavior: no item granting read access to tag <1832> at alloc906 found in borrow stack.
  --> src\main.rs:10:9
   |
10 |         dbg!(chunk1[0]);  // error here
   |         ^^^^^^^^^^^^^^^ no item granting read access to tag <1832> at alloc906 found in borrow stack.
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
```

<details>
<summary>

With `-Zmiri-track-pointer-tag=1832,1585,12319`

</summary>

```
note: tracking was triggered
    --> \library\core\src\slice\iter.rs:1564:19
     |
1564 |         Self { v: slice, chunk_size: size }
     |                   ^^^^^ created tag 1585
     |
     = note: inside `std::slice::ChunksMut::<u8>::new` at \library\core\src\slice\iter.rs:1564:19
     = note: inside `core::slice::<impl [u8]>::chunks_mut` at \library\core\src\slice\mod.rs:828:9
note: inside `main` at src\main.rs:4:20
    --> src\main.rs:4:20
     |
4    |     let mut iter = arr1.chunks_mut(8).zip(arr2.chunks(8));
     |                    ^^^^^^^^^^^^^^^^^^

note: tracking was triggered
 --> src\main.rs:5:21
  |
5 |     while let Some((chunk1, chunk2)) = iter.next() {
  |                     ^^^^^^ created tag 1832
  |
  = note: inside `main` at src\main.rs:5:21

[src\main.rs:6] chunk2[0] = 0
[src\main.rs:7] chunk1[0] = 0

note: tracking was triggered
    --> \library\core\src\slice\iter.rs:1641:32
     |
1641 |             let len = cmp::min(self.v.len().unchecked_sub(start), self.chunk_size);
     |                                ^^^^^^^^^^^^ popped tracked tag for item [Unique for <1832>] due to Read access for <1585>
     |
     = note: inside `<std::slice::ChunksMut<u8> as std::iter::Iterator>::__iterator_get_unchecked` at \library\core\src\slice\iter.rs:1641:32
     = note: inside `<std::iter::Zip<std::slice::ChunksMut<u8>, std::slice::Chunks<u8>> as std::iter::adapters::zip::ZipImpl<std::slice::ChunksMut<u8>, std::slice::Chunks<u8>>>::next` at \library\core\src\iter\adapters\zip.rs:278:23
     = note: inside `<std::iter::Zip<std::slice::ChunksMut<u8>, std::slice::Chunks<u8>> as std::iter::Iterator>::next` at \library\core\src\iter\adapters\zip.rs:84:9
    --> \library\core\src\slice\iter.rs:1642:32
     |
1642 |             from_raw_parts_mut(self.v.as_mut_ptr().add(start), len)
     |                                ^^^^^^^^^^^^^^^^^^^ created tag 12319
     |
     = note: inside `<std::slice::ChunksMut<u8> as std::iter::Iterator>::__iterator_get_unchecked` at \library\core\src\slice\iter.rs:1642:32
    --> \library\core\src\slice\mod.rs:483:5
      |
  483 | /     pub const fn as_mut_ptr(&mut self) -> *mut T {
  484 | |         self as *mut [T] as *mut T
  485 | |     }
      | |_____^ popped tracked tag for item [Disabled for <1832>] due to Write access for <12319>
      |
      = note: inside `core::slice::<impl [u8]>::as_mut_ptr` at \library\core\src\slice\mod.rs:483:5
      = note: inside `<std::slice::ChunksMut<u8> as std::iter::Iterator>::__iterator_get_unchecked` at \library\core\src\slice\iter.rs:1642:32
     note: inside `main` at src\main.rs:8:9
    --> src\main.rs:8:9
     |
8    |         iter.next();
     |         ^^^^^^^^^^^
     = note

[src\main.rs:9] chunk2[0] = 0

error: Undefined Behavior: no item granting read access to tag <1832> at alloc906 found in borrow stack.
  --> src\main.rs:10:9
   |
10 |         dbg!(chunk1[0]);  // error here
   |         ^^^^^^^^^^^^^^^ no item granting read access to tag <1832> at alloc906 found in borrow stack.
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
```

</details>

A few things to note here:

1. Access to the mutable chunk will fail. For example `arr1.chunks(8).zip(arr2.chunks_mut(8))` will fail on `chunk2[0]` instead.
2. `zip()` is required. This is probably due to `Zip` uses `__iterator_get_unchecked()` internally rather than `next()`.
3. `chunks_mut()` and `chunks_exact_mut()` can both reproduce.
4. `array_chunks_mut()` can't reproduce.

### Environment

```
$ rustc --version --verbose
rustc 1.61.0-nightly (45e2c2881 2022-02-20)
binary: rustc
commit-hash: 45e2c2881d11324d610815bfff097e25c412199e
commit-date: 2022-02-20
host: x86_64-pc-windows-msvc
release: 1.61.0-nightly
LLVM version: 14.0.0

$ cargo miri --version
miri 0.1.0 (0db4090 2022-02-12)
```

### PR DESCRIPTION
**Pull request by saethlin**
Fixes https://github.com/rust-lang/rust/issues/94231, details in that issue.
cc @RalfJung 

This isn't done just yet, all the safety comments are placeholders. But otherwise, it seems to work.

I don't really like this approach though. There's a lot of unsafe code where there wasn't before, but as far as I can tell the only other way to uphold the aliasing requirement imposed by `__iterator_get_unchecked` is to use raw slices, which I think require the same amount of unsafe code. All that would do is tie the `len` and `ptr` fields together.

Oh I just looked and I'm pretty sure that `ChunksExactMut`, `RChunksMut`, and `RChunksExactMut` also need to be patched. Even more reason to put up a draft.

### CODE REVIEW

**Comment by the8472** [comment]
> but as far as I can tell the only other way to uphold the aliasing requirement imposed by __iterator_get_unchecked is to use raw slices, which I think require the same amount of unsafe code

But that could be improved by implementing slicing and `split_at` on them.

File: library/core/src/ptr/mut_ptr.rs [hunk 1]
```diff
@@ -1262,6 +1262,103 @@ impl<T> *mut [T] {
...
+    ///
+    /// # Panics
+    ///
+    /// Panics if `mid > len`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut v = [1, 0, 3, 0, 5, 6];
+    /// let (left, right) = v.split_at_mut(2);
```
Comment (the8472) [inline library/core/src/ptr/mut_ptr.rs:1294]: The examples aren't exercising the raw pointer methods.

Comment (saethlin) [reply library/core/src/ptr/mut_ptr.rs:1294]: I'm just pasting from the normal slice methods to make things compile. This is still a draft, I think the extra raw slice methods need a tracking issue and separate PR.

**Comment by saethlin** [comment]
I think that sounds awesome. If there is someone else willing to do the new slice pointer APIs that would be ideal, my plate is pretty full already so it might take me some time to get through that whole process.

**Comment by the8472** [comment]
> If there is someone else willing to do the new slice pointer APIs that would be ideal

#95594

**Comment by saethlin** [comment]
I am stuck. The code works, but I cannot figure out how to write safety comments in most of these cases, because what makes the unsafe code valid is highly nonlocal. Basically I imagine these all as "this is safe because it is an invariant of the type" but that doesn't seem remotely helpful as a safety comment. Help?

Here are where I'm definitely missing safety comments (those that exist may need improving too):
```
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:1762: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2032: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2075: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2076: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2728: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2729: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2788: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:2789: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:3035: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:3061: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:3087: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:3106: undocumented unsafe
tidy error: /home/ben/rust/library/core/src/slice/iter.rs:3107: undocumented unsafe
```

File: library/core/src/slice/iter.rs [hunk 2]
```diff
@@ -1629,14 +1629,15 @@ unsafe impl<'a, T> TrustedRandomAccessNoCoerce for Chunks<'a, T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 #[must_use = "iterators are lazy and do nothing unless consumed"]
 pub struct ChunksMut<'a, T: 'a> {
-    v: &'a mut [T],
+    v: *mut [T],
```
Comment (the8472) [inline library/core/src/slice/iter.rs:1632]: internal overview comment here

**Comment by the8472** [comment]
> I am stuck. The code works, but I cannot figure out how to write safety comments in most of these cases, because what makes the unsafe code valid is highly nonlocal. Basically I imagine these all as "this is safe because it is an invariant of the type" but that doesn't seem remotely helpful as a safety comment. Help?

It's mostly about the new `*mut [T]::split_at_mut` calls, right?

I think this can be addressed in several steps

* put a general comment on the struct field that gives an overview
  * constructed from a valid &mut [T]
  * splitting always in-bounds, producing exclusively-owned subslices that can be converted back to `&mut [T]` and that keeps `* mut [T]::len` valid, which in turn makes the bounds checking in `split_at_mut` valid.
  * special-case: see `__iterator_get_unchecked`
* perhaps the `split_at_mut` docs can be improved? Currently state the *general* safety requirements, but perhaps they could have a better explanation in which specific circumstances those requirements are automatically fulfilled.

Or perhaps even simpler: We could explain that all uses should be treated as-if it were a `&mut [T]`, which make the whole thing safe except in the case where `__iterator_get_unchecked` is called, but then the `TrustedRandomAccess` contract takes over requiring only-once access to each subslice.

File: library/core/src/slice/iter.rs [hunk 3]
```diff
@@ -1629,6 +1629,11 @@ unsafe impl<'a, T> TrustedRandomAccessNoCoerce for Chunks<'a, T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 #[must_use = "iterators are lazy and do nothing unless consumed"]
 pub struct ChunksMut<'a, T: 'a> {
+    // This slice pointer must point at a valid region of T with at least length v.len(). Normally,
+    // those requirements would mean that we could instead use a &mut [T] here, but we cannot
+    // because __iterator_get_unchecked needs to return &mut [T], which guarantees certain aliasing
+    // properties that we cannot uphold if we hold on to the full original &mut [T]. Wrapping a raw
+    // slice instead lets us hand out non-overlapping &mut [T] subslices of the slice we wrap.
```
Comment (the8472) [inline library/core/src/slice/iter.rs:1636]: I think should be doc comment and have a `# Safety` heading. A bit unusual to put it on a field but we don't want to have it in the public API. That's assuming the docs will stay hidden due to being a private field.

And the `__iterator_get_unchecked` safety comments should be updated to refer here too.

File: library/core/src/slice/iter.rs [hunk 4]
```diff
@@ -1651,10 +1656,10 @@ impl<'a, T> Iterator for ChunksMut<'a, T> {
             None
         } else {
             let sz = cmp::min(self.v.len(), self.chunk_size);
-            // SAFETY: sz cannot exceed the slice length based on the calculation above
+            // SAFETY: This type ensures that any split_at_mut on self.v is valid.
```
Comment (the8472) [inline library/core/src/slice/iter.rs:1659]: I think `This type` is confusing if one doesn't have context. We should refer to it as something like

> The self.v contract [...]

That makes the safety section on the field a sort of reference point.

File: library/core/src/slice/iter.rs [hunk 5]
```diff
@@ -3174,7 +3220,11 @@ where
             let mut len = 1;
             let mut iter = self.slice.windows(2);
             while let Some([l, r]) = iter.next() {
-                if (self.predicate)(l, r) { len += 1 } else { break }
+                if (self.predicate)(l, r) {
+                    len += 1
+                } else {
+                    break;
+                }
```
Comment (the8472) [inline library/core/src/slice/iter.rs:3227]: accidental formatting changes? they don't look relevant to this PR

Comment (saethlin) [reply library/core/src/slice/iter.rs:3227]: Gosh dangit, I keep leaving my editor's autoformatter on and for whatever reason it keeps misformatting this file.

**Comment by RalfJung** [comment]
Is the PR description outdated? With that description (which will now live forever in our git history) it doesn't sound like this was ready to land...

**Comment by saethlin** [comment]
🤦 it's outdated

**Pull request merged**